Add support for delta basebackups [BF-6]

alexole opened the pull request:
Delta backups are based on the hashes of the files. The idea is to create a Snapshot of the data files, which results in a manifest file, describing the hashes of all the files needed to be backed up. New hashes are uploaded to the storage and used together with complementary manifest from control files for restoration.

Functionality in `rohmu` was mainly taken from `Astacus` (so it can be re-used later by Astacus) with slightly changed parts, so we can deal with hashes being changed while uploading.

To deal with situation when hash changes during the backup/upload, the file is prepared for upload with a temporary name first and then being uploaded to the name with the real hash.

deztructor commented on pghoard/rohmu/delta/snapshot.py:
```diff
@@ -0,0 +1,231 @@
+        file list contained in root object.
+
+        Note that any call to public API MUST be made with
+        snapshotter.lock held. This is because Snapshotter is process-wide
```
One option protecting against calling these methods w/o preceding `lock` being held by making these methods protected and introducing method creating context manager locking `lock` and returning facade object exposing these 3 methods.

alexole replied:
As PR description says, this part is taken from Astacus, I did not want to change the behavior and interfaces, so it can be re-used as is.

deztructor replied:
Is it possible to split the part taken from Astacus into a separate commit?

alexole replied:
I could, but it's pretty easy to split it, everything in rohmu folder is taken from Astacus almost as it is.

deztructor replied:
Sorry for me being out of the context: I wonder is it possible to split this functionality shared by Astacus and pghoard into a separate library?

alexole replied:
This is what's probably coming, was discussed multiple times, rohmu should probably move to a separate library itself.

deztructor replied:
Sounds like a good time to create a story :wink:

alexole replied:
Yes, but that's a bit out of the scope of the current ticket :) and will require quite some effort as touches many places.

deztructor replied:
Can you create a story? :blush:

alexole replied:
yep, will check first maybe we have something already.

alexole replied:
Ticket created.

deztructor commented on pghoard/rohmu/delta/snapshot.py:
```diff
@@ -0,0 +1,231 @@
+        st = src_path.stat()
+        return SnapshotFile(relative_path=relative_path, mtime_ns=st.st_mtime_ns, file_size=st.st_size)
+
+    def _get_snapshot_hash_list(self, relative_paths):
```
nit: maybe not `_get...list` but `_gen_snapshot_hashes` to make clear that it returns not `list` but `generator`?

deztructor commented on pghoard/rohmu/delta/snapshot.py:
```diff
@@ -0,0 +1,231 @@
+            src_dirs = set()
+            src_files = set()
+            for item in self.src_iterate_func():
+                if os.path.isfile(item) and not os.path.islink(item):
```
nit: what do you think about:
```
for item...:
    path = Path(item)
    if path.is_file...:
        src_files.add(item.relative_to(...
...
```
so, there will be uniform usage of `Path` along the code, as also code doesn't assume that `src_iterate_func()` generates strings (and less conversions/usage of misc. libs).

deztructor commented on pghoard/pghoard.py:
```diff
@@ -297,6 +298,36 @@ def delete_remote_basebackup(self, site, basebackup, metadata):
                             chunk["chunk_filename"],
                         )
                     )
+        elif metadata.get("format") == BaseBackupFormat.delta_v1:
```
I wonder could function be more observable if the logic for the `delta_v1` format is extracted into a separate function, so we don't have long functions with multiple levels nesting for the sake of readability.

deztructor commented on pghoard/pghoard.py:
```diff
@@ -297,6 +298,36 @@ def delete_remote_basebackup(self, site, basebackup, metadata):
+                if delta_backup["name"] == basebackup:
+                    continue
+                delta_backup_key = os.path.join(
+                    self.config["backup_sites"][site]["prefix"], "basebackup", delta_backup["name"]
```
`self.config["backup_sites"][site]["prefix"]` is used several times here, can be extracted into a variable.

deztructor commented on pghoard/pghoard.py:
```diff
@@ -297,6 +298,36 @@ def delete_remote_basebackup(self, site, basebackup, metadata):
                         )
                     )
+        elif metadata.get("format") == BaseBackupFormat.delta_v1:
+            keep_hexdigests = set()
```
Consider a test for the logic?

alexole replied:
There are tests which cover recovery functionality, they are not covering 100%, but they do cover backup and restoration, which touches pretty much every line of this code, and one test is testing the fetcher for delta data to recover. Could you tell which part exactly you want to cover?

deztructor commented on pghoard/transfer.py:
```diff
@@ -51,6 +51,7 @@ def defaults():
                 return {
                     "basebackup": EMPTY.copy(),
                     "basebackup_chunk": EMPTY.copy(),
+                    "basebackup_delta": EMPTY.copy(),
```
Is it a enum?

alexole replied:
My goal is not to refactor everything :)

alexole commented on pghoard/basebackup.py:
```diff
@@ -612,8 +773,96 @@ def create_and_upload_chunks(self, chunks, data_file_format, temp_base_dir):
+
+            uploader = Uploader(storage=delta_storage)
+            progress = Progress()
+            upload_input_size, upload_result_size, upload_files_count = uploader.write_hashes_to_storage(
```
The small files are embed into the manifest itself and are being restored here https://github.com/aiven/pghoard/blob/b068a52e8e052c2c2012afa54ec249cb72be22e3/pghoard/restore.py#L423

fingon commented on pghoard/rohmu/delta/upload.py:
```diff
@@ -0,0 +1,73 @@
+                    current_hexdigest = hash_hexdigest_readable(f)
+                if current_hexdigest != snapshotfile.hexdigest:
+                    logger.info("Hash of %s changed after upload", snapshotfile.relative_path)
+                    storage.delete_hexdigest(hexdigest)
```
There's also cases like:

- backup T has hash A
- backup T+1 does not have
- backup T+2 has hash A again

as long as any of the backups {T, T+2} want to be kept, hash A has to be kept in storage.

fingon commented on pghoard/basebackup.py:
```diff
@@ -54,6 +61,152 @@ class NoException(BaseException):
+    def list_jsons(self):
+        return {
+            os.path.basename(o["name"]): o["size"]
+            for o in self.storage.list_iter(os.path.join(self.prefix, "basebackup_delta"), with_metadata=False)
```
I'm not sure about pg file size distribution; really small files will be embedded, but medium-sized files *might* need to be stored in some shared containers. Large files are ok.

alexole replied:
Each table in one file for < 1GB (for big tables depends on the size of the table, as it's split into 1GB files). I've changed it to get existing digests from manifest files, so we don't iterate over all objects.

alexole commented on pghoard/rohmu/delta/upload.py:
```diff
@@ -0,0 +1,73 @@
+                    logger.warning("%s disappeared post-snapshot", path)
+                    continue
+                with snapshotfile.open_for_reading(snapshotter.dst) as f:
+                    current_hexdigest = hash_hexdigest_readable(f)
```
Removed redundant reads.

alexole commented on pghoard/rohmu/delta/snapshot.py:
```diff
@@ -0,0 +1,232 @@
+                continue
+            if old_snapshotfile:
+                snapshotfile.hexdigest = old_snapshotfile.hexdigest
+                snapshotfile.content_b64 = old_snapshotfile.content_b64
```
Left this logic conditional for Astacus potentially. Now I am uploading every file individually and not reusing data from previous snapshot files.

alexole commented on pghoard/basebackup.py:
```diff
@@ -54,6 +61,152 @@ class NoException(BaseException):
+
+            try:
+                # TODO Make the file persist over the with-block with this hardlink
+                os.link(raw_output_obj.name, chunk_path)
```
As for the hardlinks in delta folder I'll add removing after the backup.
The link here to the chunk_path will be deleted after upload in TransferAgent

alexole commented:
Move delta part out of `basebackup.py` into a separate file.
Removed redundant/unusable `Astacus` parts from `rohmu`.
Implemented uploading using temporary names first, so it's possible to deal with situation when hash of the file is changed during the upload.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+            "site": self.site,
+            "type": "UPLOAD",
+        })
+
```
This has no limiting for how much data may be pending for upload. If transfer speed is low this will very quickly consume all available disk space if temp files are stored on a smaller volume (as they are in our case). For regular backups we have `basebackup_chunks_in_progress` controlling the rough local disk usage, single chunk is 2 gigs (likely smaller when compressed) so local disk usage is never more than `basebackup_chunks_in_progress * 2 * 1024 * 1024 * 1024`. Instead of tracking disk usage the upload could be done directly from here as suggested in other comment, which could potentially simplify things overall (though in that case we'd have fewer upload threads than usual, which could slow down upload speed).

alexole replied:
Now the limiting is done by the uploading threads pool size.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+
+        self.log.info("Waiting for the upload of delta backup files (%r) to finish", upload_files_count)
+
+        def _post_process_delta_file(upload_result):
```
Looks like result could also be failure but I don't see any handling for that. I suppose if uploading any file failed the entire backup should be considered to have failed (or then we need to retry the failed file until it succeeds). If single failure means the entire backup failed checking results only at the very end could be quite wasteful as we could've upload a lot of files totally in vain. Also if the backup is considered to have failed we should presumably delete any files that didn't yet exist.

alexole replied:
I've changed it to the same logic we have for usual basebackups (with retries), moved to transfer.py

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+            raw_output_obj.seek(0)
+
+            os.link(raw_output_obj.name, chunk_path)
+
```
We could / should check the produced hash at this point and skip upload in case new hash matches that of any file that already exists remotely, also including files that we've processed in relation to this backup so that we don't unnecessarily upload multiple files all having the same hash.

alexole replied:
I'll add check for the current backup, but do you think it's worth to check remotely per each object? It cost money and slows down everything, also before the backup we do check which digests are already uploaded (from manifest files), so it's mainly relevant to the current backup.

alexole replied:
Un-resolve the conversation please if you still think I am missing out something (WRT current backup - added check)

rikonen replied:
I didn't mean checking remotely, just from the list of hashes we've already obtained earlier from the delta metadata files.

rikonen replied:
This is unlikely very relevant from the point-of-view of trying to optimize uploads but if proper cleanup is missing, as it looks to me, then only uploading files that weren't already there would make it easier to keep track of what we need to delete in cleanup.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+            storage_delta_path = os.path.join(self.site_config["prefix"], "basebackup_delta")
+            old_key = os.path.join(storage_delta_path, opaque["object_path"])
+            new_key = os.path.join(storage_delta_path, opaque["hexdigest"])
+            self.storage.copy_file(source_key=old_key, destination_key=new_key)
```
Will this just overwrite existing file with the same name? Just checking there won't be problems if we did new partial delta backup and pghoard was shut down before that completed and we have a bunch of remote files that aren't tracked (probably not an issue from storage usage point-of-view but just need to make sure they don't interfere with the rest of the logic).

alexole replied:
After I moved this part into `transfer.py` the upload should retry until succeeds. After upload fails, file on the disk won't be removed and possible temp object with the same name will exist as well, but will be rewritten and re-copied with the next successful retry.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+                upload_size += size
+                upload_stored_size += stored_size
+
+        self.log.info("All delta file were processed successfully")
```
As a general note calling these delta files is a bit confusing. It suggests we keep track of changes within individual files while it's only the full backups themselves that try to reuse full files of other full backups if already available. Not sure if it's worthwhile doing anything about that.

alexole replied:
Yeah, might be confusing a bit. I think I was thinking too much about `delta` lately :) 
Changed that logging and also in some other places, let me know if there is anything else confusing which I could change.

rikonen commented on pghoard/pghoard.py:
```diff
@@ -274,13 +277,47 @@ def delete_remote_wal_before(self, wal_segment, site, pg_version):
                 self.metrics.unexpected_exception(ex, where="delete_remote_wal_before")
 
-    def delete_remote_basebackup(self, site, basebackup, metadata):
+    def _get_delta_basebackup_files(self, site, storage, metadata, basebackup, keep_backups):
```
Naming could be improved to make it clearer `basebackup` is a basebackup to delete.

alexole commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+    def _list_existing_files(self):
+        """Iterate through all manifest files and fetch information about delta files"""
+        all_snapshot_files: Dict[str, SnapshotFile] = {}
+        for backup in self.get_remote_basebackups_info(self.site):
```
The prefix is built base on the site in get_remote_basebackups_info method, so if the config changes, the backups will be listed for new prefix only
https://github.com/aiven/pghoard/blob/216271d1247903261079de1c4b23c691071b4ea8/pghoard/pghoard.py#L365

alexole commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,347 @@
+        def update_hash(data):
+            uploaded_hash.update(data)
+
+        with NamedTemporaryFile(dir=temp_dir, prefix=os.path.basename(chunk_path), suffix=".tmp") as raw_output_obj:
```
I see your point, that would probably be a nice idea.
My motivation to do like this was to use already existing uploading mechanism (in transfer.py), so that I can re-use all the logic for metrics/retrying/state and no need to introduce my own, which would probably lead to maintenance difficulties.

alexole replied:
As discussed in private, I'll keep it as is.

alexole commented:
* Moved post-processing to `transfer.py` to re-use the same mechanism for retrying and error handling.
* Added limitation to amount of simultaneous files being uploaded.
* Do not upload hashes which where already submitted for uploading for the current backup.
* Some other things according to comments.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,314 @@
+        if progress.finished_successfully:
+            self.log.info("All basebackup files were uploaded successfully")
+        else:
+            raise BackupFailure("Error while uploading backup files")
```
When this fails how aggressively will it be retried? Could already be existing problem but just checking it won't end up in some very busy retry loop. It could be more relevant with this backup logic as this potentially causes a lot of calls to the cloud provider APIs while the regular tar backup mostly just loads the node locally.

alexole replied:
That's a good point, I guess it will try until succeeds, probably I need to add one metric, which should count how many backups failed and stop backup if more than X consecutive backups failed. What do you think?

rikonen replied:
Maybe some exponential backoff so that it won't start new backup until enough time has passed since last failure?

alexole replied:
Actually infinite retrying of backup would happen in case if there is no backups, added retry/backoff there. For the usual backup request it won't happen.
However there is another retry/backoff mechanism in uploading part, I am not touching it, keeping the same logic.

rikonen reviewed (dismissed):
There's a comment saying disk usage is now controlled by limiting number of threads but I don't see any change related to this nor do I understand how limiting number of threads helps if we just keep on compressing and encrypting files and putting them to upload queue without checking how many bytes we currently have pending in the upload queue.

alexole commented on pghoard/basebackup_delta.py:
```diff
@@ -0,0 +1,314 @@
+        if progress.finished_successfully:
+            self.log.info("All basebackup files were uploaded successfully")
+        else:
+            raise BackupFailure("Error while uploading backup files")
```
Probably I misunderstood it in the beginning, I was thinking about local files, which are deleted after upload (and this is done now the same way as for old backups).

So the cleanup for the current basebackups is happening when we remove old basebackups, we do the same for delta backups.
As uploading happens asynchronously in transfer agent and there is a retry logic, once the files are submitted successfully for upload, we just wait for this upload to happen. If it does not happen, it will keep on retying in the transfer agent (same logic for basebackups).
So if submission fails for any reason, then we are raising here BackupFailure and not cleaning up anything, same as logic for base backup. I could add removal of all the files which where submitted for the upload if one of the submission fails here, but this will not cover the case when one of the file keeps on failing to upload.

alexole replied:
As for the general comment regarding limiting number of files, this part is responsible for that
https://github.com/aiven/pghoard/blob/a70589917c1e92c6023c70f40a0b79fa019ef403/pghoard/basebackup_delta.py#L222
Every thread submitted something for upload, will wait for the file to be uploaded, this it's not possible to process more files than number of threads.

rikonen replied:
Ah, missed that callback_queue.get() part

alexole replied:
Added best effort cleanup of new hash files belonging to the current backup.

Ormod commented:
This is more of a meta comment but couldn't immediately see places where this would add metrics to monitor how is this doing. As in I'm interested in seeing at least metric data points on what's the savings per basebackup (because we had the data mostly there), how long did the basebackup take (i.e. because we didn't have to upload anything) and such

alexole commented:
* Added retry/back-off for backup failure
* Added best effort removal of the files which where submitted/uploaded in case of submission failure
* Added backup metrics, so we can use them to figure out the efficiency

rikonen commented on pghoard/basebackup.py:
```diff
@@ -780,11 +780,17 @@ def run_local_tar_basebackup(self, delta=False):
+                self.metrics.gauge(
+                    "pghoard.backup_time_{}".format(self.site_config["basebackup_mode"]),
+                    backup_time,
+                    tags={"site": self.site}
```
I don't think site is very interesting value as a tag. In our context it's random and not good for metrics cardinality

alexole replied:
Yep, I agree, will remote it, I've just seen it for other metrics.

rikonen commented on pghoard/basebackup_delta.py:
```diff
@@ -252,46 +252,76 @@ def _submit_files_in_thread(hexdigest):
+        self.metrics.increase("pghoard.delta_backup_total_size", inc_value=uploaded_size, tags={"site": self.site})
+        self.metrics.gauge("pghoard.delta_backup_upload_size", uploaded_size, tags={"site": self.site})
+        self.metrics.increase("pghoard.delta_backup_total_files", inc_value=uploaded_count, tags={"site": self.site})
+        self.metrics.gauge("pghoard.delta_backup_upload_files", uploaded_count, tags={"site": self.site})
```
Maybe also add number of bytes not uploaded (i.e. amount of data that remained the same). That seems to be available as percentage which may be more useful but the absolute number could be interested too.

rikonen commented on pghoard/pghoard.py:
```diff
@@ -566,6 +632,19 @@ def handle_site(self, site, site_config):
+                    return
+                else:
+                    if utc_now(
+                    ) >= self.delta_backup_failures[site].last_failed_time + datetime.timedelta(seconds=2 ** retries):
```
This ends up adding very little delay. 1 second the first time, 2 seconds the second time, and so on. I think something like 30 seconds first time, 2 minutes second, 5 minutes third, 15 minutes fourth etc. with maybe a cap at one hour would be more sensible.

rikonen reviewed (dismissed):
I'd drop the site tag everywhere as that doesn't seem useful (but can be harmful) and adjust the backoff times, otherwise looks good to me

alexole commented:
@rikonen Addressed last comments, @Ormod also wanted to take a look at the PR, if you are to approve it, could you leave in approved state without merging just yet. Thanks!
